Add introductory remix exercises

arbitrary-alex opened the pull request.

arbitrary-alex committed:
Add introductory remix exercises

pbwaffles commented on remix_exercises/Challenge0.sol:
```diff
@@ -0,0 +1,14 @@
+    event Winner(address);
+
+    function callMe() external {
+	emit Winner(msg.sender);
```
Can't seem to make a suggestion for some reason but it looks like there isnt any indentation in the function body

arbitrary-alex replied:
Weird, ok I can fix that.

pbwaffles commented on remix_exercises/Challenge0.sol:
```diff
@@ -0,0 +1,14 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+interface Challenge0 {
```
whats the point of this interface? wouldn't you just deploy the `Challenge0` contract and hit the callMe button?

arbitrary-alex replied:
I was thinking we'd deploy the contracts and have them call them using the interface, that way (as you mention below) they can read the public secret value but would have to access storage to find the private value.

pbwaffles commented on remix_exercises/Challenge0.sol:
```diff
@@ -0,0 +1,14 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
```
I think it might be cool to insert some hardhat log statements in different spots for these remix exercises. Inserting some I think will help them out later as they progress onto harder things

arbitrary-alex replied:
That makes sense, I think it depends on whether we want to deploy the challenges to Goerli for them to solve in remix or have them deploy their own instance locally.

pbwaffles commented on remix_exercises/Challenge0.sol:
```diff
@@ -0,0 +1,14 @@
+contract Challenge0 {
+    event Winner(address);
+
+    function callMe() external {
```
What do you think about adding in a `require(msg.value == 5 ether, "This function costs 5 ether!")`? This will teach them how to set the value in a transaction.

arbitrary-alex replied:
Yeah that's a good idea.

pbwaffles commented on remix_exercises/Challenge2.sol:
```diff
@@ -0,0 +1,21 @@
+    event Winner(address);
+    public uint secret;
+
+    constructor(uint _secret) {
```
do you intend to have them deploy this on goerli? if this is a local blockchain i think it seems too easy to deploy `Challenge2` with a constructor value of "10" and then just put 10 as the arg for the `callMe` function

pbwaffles replied:
(they know what they deployed the contract with so they dont have to read the public `secret` variable

arbitrary-alex committed:
Update remix exercise contracts

arbitrary-alex committed:
Fix remix exercise indentation

arbitrary-alex committed:
Fix remix exercises

arbitrary-alex committed:
Add challenge descriptions and addresses

arbitrary-alex committed:
Make remix exercises more self-contained

The idea is to have them poke around remix, deploy and manipulate contracts, not
to play a simplified CTF

arbitrary-alex committed:
Fix Challenge3 syntax
